src/OpenTK.Windowing.Desktop/Win32WindowProc.cs

```diff
@@ -0,0 +1,162 @@
+
+        // Invokes the 32- or 64-bit version depending on pointer size.
+        private static IntPtr GetWindowLongPtr(IntPtr hWnd, int nIndex)
+        {
+            if (IntPtr.Size == 8)
+            {
+                return GetWindowLongPtr64(hWnd, nIndex);
+            }
+
+            return GetWindowLongPtr32(hWnd, nIndex);
+        }
+        */
+
+        private enum WindowLongFlags : int
+        {
+            GWL_EXSTYLE = -20,
+            GWLP_HINSTANCE = -6,
+            GWLP_HWNDPARENT = -8,
+            GWL_ID = -12,
+            GWL_STYLE = -16,
+            GWL_USERDATA = -21,
+            GWL_WNDPROC = -4,
+            DWLP_USER = 0x8,
+            DWLP_MSGRESULT = 0x0,
+            DWLP_DLGPROC = 0x4,
+        }
+
+        [DllImport("user32.dll", CharSet = CharSet.Auto)]
+        private static extern IntPtr CallWindowProc(IntPtr wndProc, IntPtr hWnd, int msg, IntPtr wParam, IntPtr lParam);
+
+        private delegate IntPtr WndProcPointer(IntPtr hwnd, int msg, IntPtr wparam, IntPtr lparam);
+
+        // Indicates the beginning of modal size/move activity.
+        private const int WM_ENTERSIZEMOVE = 0x0231;
+
+        // Indicates the completion of modal size/move activity.
+        private const int WM_EXITSIZEMOVE = 0x0232;
+
+        // On older versions of Windows, this is sent when a user initiates
+        // sizing but then does not actually resize the window. In that case,
+        // WM_EXITSIZEMOVE is not sent. This can be sent in other scenarios
+        // so it is also necessary to track whether WM_ENTERSIZEMOVE has been
+        // received, in order to use this as a proxy for WM_EXITSIZEMOVE.
+        private const int WM_CAPTURECHANGED = 0x0215;
+
+        // Native win32 window handle
+        private IntPtr _hWnd = IntPtr.Zero;
+
+        // Original message-handler
+        private IntPtr _DefaultWndProc = IntPtr.Zero;
+
+        // When true, WM_ENTERSIZEMOVE has been received, and this will reset
+        // to false when WM_ENTERSIZEMOVE or WM_CAPTURECHANGED is received.
+        private bool _isModalSizeMoveState = false;
+
+        /// <summary>
+        /// Indicates the window is in a modal size/move state and update events
+        /// are temporarily suspended.
+        /// </summary>
+        public bool IsModalSizeMoveState { get => _isModalSizeMoveState; }
+
+        /// <summary>
+        /// Invoked when Windows indicates modal move/size activity has started (the user
+        /// is dragging or resizing the window).
+        /// </summary>
+        public event Action OnModalSizeMoveBegin;
+
+        /// <summary>
+        /// Invoked when Windows indicates modal move/size activity has ended.
+        /// </summary>
+        public event Action OnModalSizeMoveEnd;
+
+        public Win32WindowProc(Window* window)
+        {
+            _hWnd = GLFW.GetWin32Window(window);
+            var openTKWndProc = Marshal.GetFunctionPointerForDelegate((WndProcPointer)WndProc);
```

Comment:
A reference to the delegate created needs to be retained here, otherwise the delegate object is going to be garbage collected and the callback is going to stop working and throw an exception.

Reply:
Good catch! Thanks.